lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp

```diff
@@ -245,158 +284,173 @@ bool OwnershipModelEliminatorVisitor::visitSwitchEnumInst(
-    if (SILValue NewV = simplifyInstruction(ProjInst)) {
-      Result->replaceAllUsesWith(NewV);
-      ProjInst->eraseFromParent();
-      continue;
-    }
+    const auto &proj = projections[index];
+    SingleValueInstruction *projInst =
+        proj.createObjectProjection(builder, loc, destructureOperand).get();
 
-    Result->replaceAllUsesWith(ProjInst);
+    // First RAUW Result with ProjInst. This ensures that we have a complete IR
+    // before we perform any simplifications.
+    result->replaceAllUsesWith(projInst);
   }
 
   // Now that all of its uses have been eliminated, erase the destructure.
-  I->eraseFromParent();
+  eraseInstruction(destructureInst);
 }
 
 bool OwnershipModelEliminatorVisitor::visitDestructureStructInst(
-    DestructureStructInst *DSI) {
-  splitDestructure(B, DSI, DSI->getOperand());
+    DestructureStructInst *dsi) {
+  splitDestructure(dsi, dsi->getOperand());
   return true;
 }
 
 bool OwnershipModelEliminatorVisitor::visitDestructureTupleInst(
-    DestructureTupleInst *DTI) {
-  splitDestructure(B, DTI, DTI->getOperand());
+    DestructureTupleInst *dti) {
+  splitDestructure(dti, dti->getOperand());
   return true;
 }
 
 //===----------------------------------------------------------------------===//
 //                           Top Level Entry Point
 //===----------------------------------------------------------------------===//
 
-static bool stripOwnership(SILFunction &F) {
+static bool stripOwnership(SILFunction &func) {
   // If F is an external declaration, do not process it.
-  if (F.isExternalDeclaration())
+  if (func.isExternalDeclaration())
     return false;
 
   // Set F to have unqualified ownership.
-  F.setOwnershipEliminated();
+  func.setOwnershipEliminated();
 
-  bool MadeChange = false;
-  SILBuilder B(F);
-  OwnershipModelEliminatorVisitor Visitor(B);
+  bool madeChange = false;
+  SmallVector<SILInstruction *, 32> createdInsts;
+  SILBuilder builder(func);
+  OwnershipModelEliminatorVisitor visitor(builder);
 
-  for (auto &BB : F) {
+  for (auto &block : func) {
     // Change all arguments to have OwnershipKind::None.
-    for (auto *Arg : BB.getArguments()) {
-      Arg->setOwnershipKind(OwnershipKind::None);
+    for (auto *arg : block.getArguments()) {
+      arg->setOwnershipKind(OwnershipKind::None);
     }
 
-    for (auto II = BB.begin(), IE = BB.end(); II != IE;) {
+    for (auto ii = block.begin(), ie = block.end(); ii != ie;) {
       // Since we are going to be potentially removing instructions, we need
       // to make sure to increment our iterator before we perform any
       // visits.
-      SILInstruction *I = &*II;
-      ++II;
+      SILInstruction *inst = &*ii;
+      ++ii;
+
+      madeChange |= visitor.visit(inst);
+    }
+  }
 
-      MadeChange |= Visitor.visit(I);
+  // Once we have finished processing all instructions, we should not be
+  // consistently in non-ossa form. Now go through any instructions and simplify
```

Comment:
@atrick when I was bringing up ossa part of my goal was zero diff (or as small as possible diff) after ownership lowering of our tests. With that in mind, I started to notice that a bunch of diff can from not running inst simplify's simplifications on split up destructures. It was completely a pragmatic decision.

That being said, OME is not an OSSA pass or a non-OSSA pass. It is more of a specialized transform that is in a between world while it is running. I don't think that is an assumption we want any of our utility code to have to deal with. Code should always be able to expect either fully non-ossa or ossa.

So it makes sense to hoist out this simplification code outside of the lowering so that it knows it structurally has non-ossa SIL. The actual thing we are simplifying are struct_extract from splitting destructures during the OME transform. So the simplification has to run after the transform on non-OSSA.

Reply:
I think I improved the wording.